proposals/p2875.md

```diff
@@ -0,0 +1,559 @@
+
+### Overloading
+
+We intend to support function overloading, and although it outside the scope of
+this proposal, it is important to ensure that we have reasonable paths forward
+to support overloading within this framework.
+
+An overloaded function has multiple different signatures -- sets of implicit and
+explicit parameters -- that will be checked against the provided arguments. In
+this proposal, that means that a set of overloaded functions introduces a single
+function type that supports being called in different ways.
+
+Depending on how we approach overloading, overload selection might be done by
+checking each candidate in turn until we find one that matches, or checking them
+all and somehow selecting the best match, and our current intent is to use the
+former approach. Translated into this proposal, one important consequence is
+that the function type corresponding to the set of overloaded functions has
+multiple parameterized implementations of the `Call(...)` interface, and the
+implementation selected for a particular call will need to follow whatever rules
+the overloading mechanism picks. For example, if function overloading picks the
+first matching function, then given the following, using placeholder syntax:
+
+```
+overloaded fn Abs[T:! Unsigned](x: T) -> T { return x; }
+overloaded fn Abs[T:! Floating](x: T) -> T { return x < 0 ? -x : x }
+```
+
+we would synthesize implementations that behave as follows:
+
+```
+match_first {
+  impl forall [T:! Unsigned] Abs as Call((T,)) where .Result = T { ... }
+  impl forall [T:! Floating] Abs as Call((T,)) where .Result = T { ... }
+}
+```
+
+For a type that is `Unsigned & Floating`, the former `impl` will be selected,
+matching the behavior of overload selection.
+
+### Overloading on expression category and phase
+
+We should consider whether calls can be overloaded on the expression category
+and expression phase of the callee and arguments. For compile-time arguments, we
+may also wish to support overloading on the constant value.
+
+Other languages support overloading on the expression category of the callable
+object. Rust provides separate `Fn`, `FnMut`, and `FnOnce` to distinguish
+between different ways of passing the callable object into an overloaded
+function call, and C++ permits `operator()` to take `*this` as `const`,
+non-`const`, or even `&&`.
+
+In Carbon, we could follow the Rust approach and provide multiple `Call`
+interfaces to handle different kinds of `self` parameter, modeled in a similar
+way to the `IndexWith` and `IndirectIndexWith` interfaces used for subscripting.
+A more general, ambitious and ergonomic approach that we have started exploring
+would be to allow the function call signature to be described as part of the
+call interface:
+
+```carbon
+impl MyCallable as call(T:! type, x: T) -> T;
+```
+
+This might be a shorthand syntax for some way of expressing the signature as an
+interface, or a new form of first-class language primitive.
```

Comment:
Is there a good place to link to the issue where we're tracking the longer-term effort here?

Reply:
Done.